docs/contributor/writing-effective-tests.md

```diff
@@ -0,0 +1,316 @@
+
+There are plans to add a third set of smoke tests that will have no mocking at all and actually click through the VS Code user interface to perform actions. For more information see the [Swift for Visual Studio Code test strategy](./test-strategy.md) document.
+
+## Running the Tests
+
+NPM scripts exist to run both unit and integration tests. In your terminal you can run `npm run unit-tests` to execute the unit tests or `npm run integration-test` to run the integration tests. These can also be launched with the `--coverage` flag to display converage information. For example:
+
+```bash
+npm run unit-test -- --coverage
+```
+
+If any test fails, Mocha will provide detailed information to help you debug the issue.
+
+## Writing Unit Tests
+
+Unit tests should be organized in a way that reflects the structure of the project. For example, let's say you were writing unit tests for the [`showReloadExtensionNotification()`](../../src/ui/ReloadExtension.ts) function under `src/ui/ReloadExtensions.ts`. You would want to first create a unit test file that mirrors the structure of that feature’s implementation:
+
+1. **Create a new test file** `test/unit-tests/ui/ReloadExtensions.ts` that will contain the test suite:
+
+    ```
+    test/unit-tests/ui/
+      |- ReloadExtensions.test.ts
+    ```
+
+2. **Structure your test suite** using Mocha’s `suite()` function. This function allows you to group related tests together logically.
+
+    ```typescript
+    suite("ReloadExtension Unit Test Suite", function () {
+        // Individual test cases will go here
+    });
+    ```
+
+3. **Create your first test case** using Mocha's `test()` function. Your test name should clearly explain what the test is trying to verify.
+    ```typescript
+    suite("ReloadExtension Unit Test Suite", () => {
+        test("displays a warning message asking the user to reload the window", async () => {
+            // Test code goes here
+        });
+    });
+    ```
+
+Now comes the fun part! For unit testing we use [Chai](https://chaijs.com) for assertions and [Sinon](https://sinonjs.org/) for stubbing/mocking user interaction and observing behaviour that is otherwise difficult to observe. Many helpful utility functions exist in [MockUtils.ts](../../test/MockUtils.ts). These utility methods take care of the setup and teardown of the mocks so the developer does not need to remember to do this for each suite/test.
+
+Our test is going to want to verify that a warning message is shown to the user. For this, we'll want to mock out VSCodes global `window` module. Thankfully, MockUtils contains a `mockGlobalObject` function that will replace a global object with a mocked version. Each property is replaced with a Sinon stub during setup and restored at teardown. Keep in mind that the `mockGlobal*` functions all need to be created at the **`suite()`** level. Unexpected behavior will occur if called within a `test()` function.
```

Comment:
maybe make clear the strength of this comes from the test utilities handling the setup and teardown

Reply:
I added more info on this in the `Mocking global modules` section.